services/repository/generate.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repository
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification"
+)
+
+// GenerateRepository generates a repository from a template
+func GenerateRepository(doer, owner *models.User, templateRepo *models.Repository, opts models.GenerateRepoOptions) (_ *models.Repository, err error) {
+	var generateRepo *models.Repository
+	if err = models.WithTx(func(ctx models.DBContext) error {
+		generateRepo, err = models.GenerateRepository(ctx, doer, owner, templateRepo, opts)
+		if err != nil {
+			if generateRepo != nil {
+				if errDelete := models.DeleteRepository(doer, owner.ID, generateRepo.ID); errDelete != nil {
```

Comment:
I think this should be done after `models.WithTx` finished.

Reply:
> 
> 
> I think this should be done after `models.WithTx` finished.

😳 It wasn't??

Reply:
It should be in line 59, but not 20.

Reply:
Oh, I see. The delete must happen outside the failed transaction. 👍

Reply:
Okay, I moved it into the `err != nil` after `WithTx`